# PANDAS CAPSTONE PROJECT- SCHOOL SAFETY

![schoolsafety.png](schoolsafety.png)

* Since 1998, the New York City Police Department (NYPD) has been tasked with the collection and maintenance of crime data for incidents that occur in New York City public schools. For presentation purposes, each incident has been classified in one of three categories. These categories are:
<br>
**Major Crimes:** This category is consistent with those regularly and publicly reported by the NYPD. It includes the most serious personal and property crimes. The property crimes are burglary, grand larceny and grand larceny auto. The crimes against persons are murder, rape, robbery and felony assault.
<br>
**Other Crimes:** This category is composed of many crimes and incidents that range in severity. It includes reports of incidents such as arson/explosion, misdemeanor assault, criminal possession or sale of a controlled substance, sale of marijuana, criminal mischief, petit larceny, reckless endangerment, sex offenses (not including rape, which is included in the Major Crimes), and weapons possession.
<br>
**Non-Criminal Incidents:** This category includes actions which are not classified as crimes but are nevertheless disruptive to the school environment. It includes disorderly conduct, harassment, loitering, possession of marijuana, dangerous instruments and trespass.
<br>
NYPD and NYC Department of Education are stored this crime data as annualy school safety reports and published on https://www.data.gov/ . <br>
 __In this Data Analysis exercise, I concatenated the School Safety Reports of the 2015 and 2016, and I will try to Analyse this data.__ <br>


### IMPORTING LIBRARIES
* Import `numpy` as `np` and `pandas` as `pd`

In [77]:
import numpy as np
import pandas as pd

### STEP 1: EXAMINING DF 

#### IMPORTING DATA

* Import the school safety data and name it as:
    * 2015_16ss: `ss1516`
    * 2016_17ss: `ss1617`
* Don't forget to set `encoding="utf-8"` , `quotechar='""'`, and `delimiter=","`

In [20]:
ss1516=pd.read_csv("2015_16ss.csv",encoding="utf-8",delimiter=",")
ss1617=pd.read_csv("2016_17ss.csv",encoding="utf-8",delimiter=",")

* Create `ss1517` by concating `ss1516` and `ss1617`
* Use `shape` to figure out how many rows and columns our `ss1517` has.

In [27]:
ss1517=pd.concat([ss1516,ss1617],join="inner",axis=0)

* Print the first 3 rows of `ss1517`

In [101]:
ss1517.shape

(4116, 20)

In [28]:
ss1517.head(3)

Location Name Location Code Borough  \
0           P.S. 001 The Bergen          K001       K   
1  Parkside Preparatory Academy          K002       K   
2    EXPLORE CHARTER SCHOOL(BS)          K704       K   

   Geographical District Code  Register  \
0                        15.0    1280.0   
1                        17.0     475.0   
2                        17.0     529.0   

                               Building Name  # Schools  \
0                                        NaN        1.0   
1  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        3.0   
2  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        3.0   

                                 Schools in Building  Major N  Oth N  \
0                                P.S. 001 The Bergen      0.0    1.0   
1  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   
2  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   

   NoCrim N  Prop N  Vio N ENGroupA     RangeA  AvgOfMajor N  AvgOfOth N  \
0       0.0     1.0    0.0       7C  1251-1500          0.64        3.02   
1       NaN     NaN    NaN       3C    251-500           NaN         NaN   
2       NaN     NaN    NaN       4C    501-750           NaN         NaN   

   AvgOfNoCrim N  AvgOfProp N  AvgOfVio N  
0           5.77         1.72        1.54  
1            NaN          NaN         NaN  
2            NaN          NaN         NaN

#### Explanation of the Columns is needed to understand our analysis better <br>
* __Location Name__ is the the name by which the organization is known. For a learning community, it is the official title of the school. <br>
* __Location Code__ is a unique identifier that can include schools, administrative offices, learning communities, etc. <br>
* __Borough__ is the NYC Boro the location is situated in. <br>
* __Geographical District Code__ the school’s geographical district as defined by the NYC Department of Education. <br>
* __Register__ Number of students on register. <br>
* __Building Name__ is the the official name of the building a school is located in. <br>
* __# Schools__ is the number of schools in in the building. <br>
* __Schools in the Building__ is the names of the schools in the buildings. <br>
* __Major N__ is the number of major crimes. <br>
* __Oth N__ is the number of other crimes. <br>
* __NoCrim N__ is the number of non - criminal crimes. <br>
* __Prop N__ is the number of property crimes. <br>
* __Vio N__ is the number of violent crimes. <br>
* __EnGroup A__ is the building population. <br>
* __Range A__ is the group name  building population falls under. <br>
* __AvgofMajorN__ is the average of major crimes for all buildings that have the same EnGroupA/Range A. <br>
* __AvgofOthN__ is the average of other crimes for all buildings that have the same EnGroupA/Range A. <br>
* __AvgofNoCrimN__ is the average of non-criminal crimes for all buildings that have the same EnGroupA/Range A. <br>
* __AvgofPropN__ is the average of property crimes for all buildings that have the same EnGroupA/Range A. <br>
* __AvgofVioN__ is the average of violent crimes for all buildings that have the same EnGroupA/Range A. <br>
---
Let's take a breif look of our data.

* Use `.info()` to get summary information about the `ss1517`

In [29]:
ss1517.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4116 entries, 0 to 2045
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Location Name               4115 non-null   object 
 1   Location Code               3790 non-null   object 
 2   Borough                     4113 non-null   object 
 3   Geographical District Code  4112 non-null   float64
 4   Register                    4046 non-null   float64
 5   Building Name               2373 non-null   object 
 6   # Schools                   4115 non-null   float64
 7   Schools in Building         4115 non-null   object 
 8   Major N                     2401 non-null   float64
 9   Oth N                       2401 non-null   float64
 10  NoCrim N                    2401 non-null   float64
 11  Prop N                      2401 non-null   float64
 12  Vio N                       2401 non-null   float64
 13  ENGroupA                    4073 

While 8 of our column's dtpyes are object, 12 of them's dtypes are float.
___

Let's examine the summary statistics of our ss1517 df:

* Use `.describe()` to receive summary statistics about `ss1517`

In [30]:
ss1517.describe()

Geographical District Code     Register    # Schools      Major N  \
count                 4112.000000  4046.000000  4115.000000  2401.000000   
mean                    15.419018   687.297084     2.170595     0.426489   
std                      9.221523   547.373869     1.452559     0.873994   
min                      0.000000    25.000000     1.000000     0.000000   
25%                      8.000000   354.000000     1.000000     0.000000   
50%                     14.000000   529.500000     2.000000     0.000000   
75%                     24.000000   850.000000     3.000000     1.000000   
max                     32.000000  5682.000000     8.000000     8.000000   

             Oth N     NoCrim N       Prop N        Vio N  AvgOfMajor N  \
count  2401.000000  2401.000000  2401.000000  2401.000000   2400.000000   
mean      1.746772     3.489379     1.052895     0.875052      0.438704   
std       2.849710     7.013974     1.750748     1.689985      0.347415   
min       0.000000     0.000000     0.000000     0.000000      0.240000   
25%       0.000000     0.000000     0.000000     0.000000      0.270000   
50%       1.000000     1.000000     0.000000     0.000000      0.280000   
75%       2.000000     4.000000     1.000000     1.000000      0.570000   
max      25.000000    88.000000    16.000000    13.000000      3.140000   

        AvgOfOth N  AvgOfNoCrim N  AvgOfProp N   AvgOfVio N  
count  2400.000000    2399.000000  2399.000000  2400.000000  
mean      1.836046       3.384894     1.120796     0.879746  
std       1.669645       4.018752     0.898922     0.761603  
min       0.710000       1.670000     0.510000     0.320000  
25%       0.900000       1.710000     0.630000     0.440000  
50%       1.110000       1.740000     0.750000     0.530000  
75%       2.120000       3.640000     1.390000     1.050000  
max      13.000000      34.860000     8.290000     5.290000

### STEP 2: LOCATING & REMOVING NA VALUES

Let's check whether our df has Na values or not:
* Use `isnull().values.any()` for this purpose

In [35]:
ss1517.isnull().values.any()

True

Appearently, we have some Na values. Let's figure out how many Na values we have:

* Use `isnull().values.sum()` to see how many NA values we have in `ss1517`

In [38]:
ss1517.isnull().values.sum()

19392

19392 of our values are Na. Wow!, that's a lot. In that case, we have 3 options: 
<br>
1) We can get rid of them with `ss1517.dropna()` . We can do this but we also loose a lot of useful information because `.dropna()`__drops entire column that has Na value__ , and not every value in that column is Na. That's why we won't go with dropna().
<br>
2) We can use the `value` parameter of the `fillna()` function. In that case, we can only replace Na 's with one variable. If we try to replace them with int or float, we 'll also replace the Na values in the object columns, and the reverse is also applicable. It seems little messy. <br>
3) We can use the `method`parameter of the `fillna()`function. We can set `method='ffill'`to replace Na values with the last valid observation, or we can set `method = 'bfill'`to replace Na values with the next valid observation. If we first set `method = 'ffill'`and then set `method = 'bfill'`, we can remove all Na values with the same dtype of columns. That way we are able to protect our dataframe's structure. That's why we'll go with this option. 

* Fill the NA values or drop them with the relevant method(s). Briefly explain why do you choose the particular method or why don't you choose the others..

In [47]:
ss1517=ss1517.fillna(method="ffill")

In [50]:
ss1517=ss1517.fillna(method="bfill")

Na values check, once more..

* Recheck whether you have NA values in `ss1517` or not with `isnull().values.any()`

In [52]:
ss1517.isnull().values.any()

False

* Take a look at your data with `.info()` and evaluate your data within `ss1517`

Great ! We don't have any Na values. 
___

Let's check whether we made any changes after we implement `fillna()`
* Use `info` for that purpose

In [54]:
ss1517.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4116 entries, 0 to 2045
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Location Name               4116 non-null   object 
 1   Location Code               4116 non-null   object 
 2   Borough                     4116 non-null   object 
 3   Geographical District Code  4116 non-null   float64
 4   Register                    4116 non-null   float64
 5   Building Name               4116 non-null   object 
 6   # Schools                   4116 non-null   float64
 7   Schools in Building         4116 non-null   object 
 8   Major N                     4116 non-null   float64
 9   Oth N                       4116 non-null   float64
 10  NoCrim N                    4116 non-null   float64
 11  Prop N                      4116 non-null   float64
 12  Vio N                       4116 non-null   float64
 13  ENGroupA                    4116 

Great ! Everything seems to be in order.
* Print the first 3 rows of your df.

In [104]:
ss1517.head(30
        )

Location Name Location Code Borough  \
0                                 P.S. 001 The Bergen          K001       K   
1                        Parkside Preparatory Academy          K002       K   
2                          EXPLORE CHARTER SCHOOL(BS)          K704       K   
3                                           P.S. K141          K141       K   
4           655 PARKSIDE AVENUE CONSOLIDATED LOCATION          K141       K   
5                        P.S. 003 The Bedford Village          K003       K   
6                          P.S. 005 Dr. Ronald Mcnair          K005       K   
7                                            P.S. 006          K006       K   
8                            P.S. 007 Abraham Lincoln          K007       K   
9                              P.S. 008 Robert Fulton          K008       K   
10                          P.S. 009 Teunis G. Bergen          K009       K   
11        BROOKLYN EAST COLLEGIATE CHARTER SCHOOL(BN)          K780       K   
12          80 UNDERHILL AVENUE CONSOLIDATED LOCATION          K780       K   
13  Magnet School of Math, Science and Design Tech...          K010       K   
14                           P.S. 011 Purvis J. Behan          K011       K   
15                   Dr. Jacqueline Peek-Davis School          K012       K   
16                  Ronald Edmonds Learning Center II          K484       K   
17            430 HOWARD AVENUE CONSOLIDATED LOCATION          K484       K   
18                          P.S. 013 Roberto Clemente          K013       K   
19     Achievement First East New York Charter School          K358       K   
20      557 PENNSYLVANIA AVENUE CONSOLIDATED LOCATION          K358       K   
21                              J.H.S. 014 Shell Bank          K014       K   
22                                    PS 77(BROOKLYN)          K077       K   
23       2424 BATCHELDER STREET CONSOLIDATED LOCATION          K077       K   
24                           P.S. 015 Patrick F. Daly          K015       K   
25                            P.S. 016 Leonard Dunkly          K016       K   
26   WILLIAMSBURG COLLEGIATE CHARTER SCHOOL(BROOKLYN)          K355       K   
27            157 WILSON STREET CONSOLIDATED LOCATION          K355       K   
28                        P.S. 017 Henry D. Woodworth          K017       K   
29                       Conselyea Preparatory School          K577       K   

    Geographical District Code  Register  \
0                         15.0    1280.0   
1                         17.0     475.0   
2                         17.0     529.0   
3                         17.0     374.0   
4                         17.0    1378.0   
5                         13.0     450.0   
6                         16.0     231.0   
7                         17.0     712.0   
8                         19.0     969.0   
9                         13.0     924.0   
10                        13.0     838.0   
11                        13.0     388.0   
12                        13.0     838.0   
13                        15.0     942.0   
14                        13.0     831.0   
15                        23.0     234.0   
16                        23.0     141.0   
17                        23.0     234.0   
18                        19.0     519.0   
19                        19.0     846.0   
20                        19.0     519.0   
21                        22.0     548.0   
22                        22.0     289.0   
23                        22.0     837.0   
24                        15.0     457.0   
25                        14.0     242.0   
26                        14.0     538.0   
27                        14.0     242.0   
28                        14.0     292.0   
29                        14.0     506.0   

                                    Building Name  # Schools  \
0       655 PARKSIDE AVENUE CONSOLIDATED LOCATION        1.0   
1       655 PARKSIDE AVENUE CONSOLIDATED LOCATION        3.0   
2       655 PARKSIDE AVENUE CONSO

How beautiful does our data sets look like without Na values..

### STEP 3: ANALYSING DATA

__In this dataset, major crimes are coded like this:__<br>
Burglary - 0<br>
Grand larceny - 1<br>
Grand larceny auto - 2<br>
Murder - 3<br>
Rape - 4 <br>
Robbery - 5<br>
Felony - 6<br>
Assault - 8<br>
___
Let's check it out that how is major crimes' distribution in the number of Major Crimes , a.k.a. `Major N`
* For that purpose, you can use `unique()` method.
* Name your result as `avgmajor`

In [107]:
pd.unique(ss1517["Major N"])


array([0., 3., 1., 2., 4., 6., 5., 8.])

In [110]:
for i in pd.unique(ss1517["Major N"]):
    print("Major N code {} sayısı: {}".format(i,ss1517[ss1517["Major N"]==i]["Major N"].count()))

Major N code 0.0 sayısı: 2906
Major N code 3.0 sayısı: 125
Major N code 1.0 sayısı: 700
Major N code 2.0 sayısı: 279
Major N code 4.0 sayısı: 74
Major N code 6.0 sayısı: 6
Major N code 5.0 sayısı: 25
Major N code 8.0 sayısı: 1


Though we have 7 different crimes, Burglary(0) and Grand Larcery(1) are so dominant that other crimes couldn't show themselves on the `Major N` .This explanation will be understood better once we plot our `Major N`(Number of Major Crimes) column.
___

Let's separate these two and dive deeper. and name it as `bigtwo` <br>
Hint: You can use the following structure: `df[df["Column"] <= yourfilter]`

### BIG TWO

In [85]:
bigtwo = ss1517.loc[(ss1517["Major N"] == 0.0) | (ss1517["Major N"] == 1.0)]

Let's examine our big two.
* Start with `shape`, then proceed with `info()`
* You can also print the first five columns for that purpose.

In [86]:
bigtwo.shape

(3606, 20)

In [88]:
bigtwo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3606 entries, 0 to 2045
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Location Name               3606 non-null   object 
 1   Location Code               3606 non-null   object 
 2   Borough                     3606 non-null   object 
 3   Geographical District Code  3606 non-null   float64
 4   Register                    3606 non-null   float64
 5   Building Name               3606 non-null   object 
 6   # Schools                   3606 non-null   float64
 7   Schools in Building         3606 non-null   object 
 8   Major N                     3606 non-null   float64
 9   Oth N                       3606 non-null   float64
 10  NoCrim N                    3606 non-null   float64
 11  Prop N                      3606 non-null   float64
 12  Vio N                       3606 non-null   float64
 13  ENGroupA                    3606 

There might be relevance between the tow big crimes and the Borough that they are commited. 
___
* To figure out relevance, we need to know our Borough values. Let's find the unique values in Borough column by `unique()`function. <br>
* Use `unique()` and return the unique values of `Borough` column of `bigtwo`


In [89]:
pd.unique(bigtwo.Borough)

array(['K', 'O', 'M', 'Q', 'R', 'X'], dtype=object)

In that case : <br>
__M__ represents Manhattan. <br>
__Q__ represents Queens. <br>
__R__ represents Rikers Island. <br>
__K__ represents Brooklyn. <br>
__X__ represents The Bronx. <br>
__O__ represents Staten Island. <br>

* Now, as we know Borough's and their actual names, we can examine the relationship between the bigtwo and Boroughs. We can do this through grouping them by Borough's. In that case, the most handy tool is  `groupby()`function.
* Use `groupby()` and group `bigtwo` by it's `Borough`

In [100]:
bigtwo.groupby(["Borough","Major N"]).count()

Location Name  Location Code  Geographical District Code  \
Borough Major N                                                             
K       0.0                950            950                         950   
        1.0                233            233                         233   
M       0.0                552            552                         552   
        1.0                161            161                         161   
O       0.0                  2              2                           2   
Q       0.0                644            644                         644   
        1.0                 87             87                          87   
R       0.0                142            142                         142   
        1.0                 18             18                          18   
X       0.0                616            616                         616   
        1.0                201            201                         201   

                 Register  Building Name  # Schools  Schools in Building  \
Borough Major N                                                            
K       0.0           950            950        950                  950   
        1.0           233            233        233                  233   
M       0.0           552            552        552                  552   
        1.0           161            161        161                  161   
O       0.0             2              2          2                    2   
Q       0.0           644            644        644                  644   
        1.0            87             87         87                   87   
R       0.0           142            142        142                  142   
        1.0            18             18         18                   18   
X       0.0           616            616        616                  616   
        1.0           201            201        201                  201   

                 Oth N  NoCrim N  Prop N  Vio N  ENGroupA  RangeA  \
Borough Major N                                                     
K       0.0        950       950     950    950       950     950   
        1.0        233       233     233    233       233     233   
M       0.0        552       552     552    552       552     552   
        1.0        161       161     161    161       161     161   
O       0.0          2         2       2      2         2       2   
Q       0.0        644       644     644    644       644     644   
        1.0         87        87      87     87        87      87   
R       0.0        142       142     142    142       142     142   
        1.0         18        18      18     18        18      18   
X       0.0        616       616     616    616       616     616   
        1.0        201       201     201    201       201     201   

                 AvgOfMajor N  AvgOfOth N  AvgOfNoCrim N  AvgOfProp N  \
Borough Major N                                                         
K       0.0               950         950            950          950   
        1.0               233         233            233          233   
M       0.0               552         552            552          552   
        1.0               161         161            161          161   
O       0.0                 2           2              2            2   
Q       0.0               644         644            644          644   
        1.0                87          87             87           87   
R       0.0               142         142            142          142   
        1.0                18          18             18           18   
X       0.0               616         616            616          616   
        1.0               201         201            201          201   

                 AvgOfVio N  
Borough Major N              
K       0.0             950  
        1.0             233  
M       0.0             552  
        1.0             161 

* Evaluate the result **with your own words**. Remember, Data Science is all about explaining the story of the data. So, try your best and seeze the story behind the data..

Well, it seems that burglary and grand larceny crimes are commited in Brooklyn first place, then the Bronx, then Manhattan. We can say that burglary and grand larceny crimes in NYC Public Schools are not relevant with the wealth because the richest borough of the NYC is Manhattan; and there are more burglary and grand larceny crimes commited in Manhattan than the poorest Borough in NYC, the Bronx. !!!!
___
Hmm, how about adding a new variable in our equasion and looking from different perspective ? <br>
Let's examine our big two by the Borough's and the number of students in each schools, a.k.a. `Register` <br>
That way, we can evaluate our bigtwo not only with Borough's and the wealth of them, but also with the population's of Borough's and their effects on burglary and grand larceny.
___


Now, we can say that there are much more grand larceny crime is commited than burglary in Manhattan. For other Borough's , everything is pretty much same. !!!

### OTHER MAJOR CRIMES

Now, it is time to talk about the other major crimes : <br>
Grand larceny auto - 2<br>
Murder - 3<br>
Rape - 4 <br>
Robbery - 5<br>
Felony - 6<br>
Assault - 8<br>
___
As they did not occur as frequent as the burglary and the grand larceny, their contest is much more serious than bigtwo. Let's create a new dataframe and name it as `othercrimes`:
* Hint: You can use the following structure: `df[df["Column"] <= yourfilter]`

Let's examine the other crimes: 
* Start with `shape`, then proceed with `info()`
* You can also print the first five columns for that purpose.

* Print the first 3 rows of `othercrimes` df. 

It seems that we won't use some of the columns in our dataset. Let's get rid of them. <br>
We can do this by `drop()`function of Pandas. With `drop`we can drop columns.
___
* After careful review, your team lead thinks that `Location Code` column is unnecessary for our analysis.
* Drop the `Location Code` column from `othercrimes` df.
* After that, print the first 5 rows to check whether `Location Code` is dropped or not. 

Great ! In order to make smooother analysis, we'll only need 3 columns : <br>
`# Schools` : Number of schools in the building. We are going to need this column because different schools means different population categories such as age and culture. Different categories might be the major crimes in school. <br>
`EnGroupA`: Building population. <br>
`Major N`: Other major crimes.

To study with only these 3 columns, we need to reshape our df.We can do this by `loc`and `iloc`
* Update the `othercrimes` as it has only `# Schools`, `EnGroupA`, and `Major N` columns, nothing more. Use `loc` or `iloc` for this purpose.

* Print the first 5 rows of your df's updated version.

* Let's sort the values by Major N codes to see the relevance better.
* Use `sort_values()` for that purpose.

* Great! To see the effects of Building Population and Number of Other Schools in the Building on Other Major Crimes, let's pivot it. We can pivot it by `pivot_table`function of pandas.
* Use `pivot_table` as follows:
    * For index, use `EnGroupA`,
    * For columns, use `# Schools`
    * For values, use `Major N`

We have Na values, again.. Replace them as we did earlier, with `bfill`and `ffill`parameters of `pd.fillna()`

Print the first 10 columns of `pivot_table` **df**.

Great ! In order to make smooother analysis, we'll only need 3 columns : <br>
`# Schools` : Number of schools in the building. We are going to need this column because different schools means different population categories such as age and culture. Different categories might be the major crimes in school. <br>
`EnGroupA`: Building population. <br>
`Major N`: Other major crimes.

* Evaluate your `pivot_table` df and final work with your own words.